lib/Optimizer/CodeGen/ConvertToQIR.cpp

```diff
@@ -1077,6 +1077,164 @@ class MyToMz : public OpConversionPattern<quake::MyOp> {
   }
 };
 
+/// Convert quake.unitary to a call to __quantum__qis__unitary(data,
+/// controls, targets) or __quantum__qis__constant_unitary(real, imag, controls,
+/// targets)
+class UnitaryOpRewrite : public OpConversionPattern<quake::UnitaryOp> {
+public:
+  using OpConversionPattern::OpConversionPattern;
+
+  Value mapUnitaryDataToLLVMPointer(Location &loc,
+                                    ConversionPatternRewriter &rewriter,
+                                    ModuleOp parentModule,
+                                    std::int64_t numUnitaryElements,
+                                    std::string &operationName,
+                                    ArrayRef<double> data) const {
+    auto f64Ty = rewriter.getF64Type();
+    SmallVector<std::int64_t> shape{numUnitaryElements};
+    auto tensorTy = VectorType::get(shape, f64Ty);
+    auto insertPoint = rewriter.saveInsertionPoint();
+    rewriter.setInsertionPointToStart(parentModule.getBody());
+
+    LLVM::GlobalOp realGlobalOp;
+    if (auto existingArr =
+            parentModule.lookupSymbol<LLVM::GlobalOp>(operationName))
+      realGlobalOp = existingArr;
+    else
+      realGlobalOp = rewriter.create<LLVM::GlobalOp>(
```

Comment:
I'm thinking this global should be created early and UnitaryOp should just carry a symbol name rather than a copy of the flattened matrix. Also, if it is a symbol the values in the matrix don't even need to be known at compile-time.